### DISCUSSION TITLE
Does the `date_format` validation rule support ISO weeks?

### DISCUSSION THREAD

**Discussion by jameswilddev** [laravel/framework#51357] [Q&A]
The docs state: https://laravel.com/docs/11.x/validation#rule-date-format

> The field under validation must match one of the given formats. You should use either date or date_format when validating a field, not both. This validation rule supports **all** formats supported by PHP's [DateTime](https://www.php.net/manual/en/class.datetime.php) class.

Formats in question: https://www.php.net/manual/en/datetime.format.php

These are my rules:

```
        return [
            'date' => [
                'required',
                'date_format:o-W',
            ],
        ];
```

Here's my request (sans cookies, referrer, host, etc.):

```
curl 'http://site.test/endpoint?date=2024-19' \
-X 'GET' \
-H 'Accept: application/json' \
-H 'Pragma: no-cache' \
-H 'Cache-Control: no-cache' \
-H 'Accept-Language: en-GB,en;q=0.9' \
-H 'Accept-Encoding: gzip, deflate' \
-H 'Connection: keep-alive'
```

And the corresponding response:

```
{
    "message": "The date field must match the format o-W.",
    "errors": {
        "date": [
            "The date field must match the format o-W."
        ]
    }
}
```

But PHP generates this exact string for the current date:

```
> now()->format('o-W')
= "2024-19"
```

Is this unsupported functionality (i.e. the docs are wrong) or am I doing something wrong?

**Comment by ghost** [comment]
Y-W is the format you are seeking.

**Reply by jameswilddev** [reply]
Should've mentioned that I tried that one first, had the same result.

**Reply by ghost** [reply]
`$date = DateTime::createFromFormat('!'.$format, $value);`
from \Illuminate\Validation\Concerns\ValidatesAttributes::validateDateFormat

 returns false for Y-W. I am not sure why is that ! there.
Solution is to create a custom validation rule for this situation.

createFromFormat is executed from Carbon FYI

dd(date_parse("2024-48"));

array:12 [
  "year" => 2024
  "month" => 4
  "day" => 1
  "hour" => false
  "minute" => false
  "second" => false
  "fraction" => false
  "warning_count" => 0
  "warnings" => []
  "error_count" => 1
  "errors" => array:1 [
    6 => "Unexpected character"
  ]
  "is_localtime" => false
]

**Comment by ghost** [comment]
```
    public function passes($attribute, $value): bool
    {
        try {
            $exploded = (array)\explode('-', $value);

            return \count($exploded) === 2
                && \is_numeric($y = \reset($exploded))
                && \is_numeric($w = \next($exploded))
                && Carbon::now()->setYear((int)$y)->week((int)$w)->format('Y-W') === $value;
        } catch (\Throwable $e) {
            return false;
        }
    }
```
